Update semantics doc for streaming writes (#3079)
GoogleCloudPlatform/gcsfuse: docs/semantics.md

until they are written out to Cloud Storage, you
must ensure that there is enough free space available to handle staged content
when writing large files.

#### Notes

- Prior to version 1.2.0, you will notice that an empty file is created in the
  Cloud Storage bucket as a hold. Upon closing or fsyncing the file, the file
  is written to your Cloud Storage bucket, with the existing empty file now
  reflecting the accurate file size and content. Starting with version 1.2,
  the default behavior is to not create this zero-byte file, which increases
  write performance. If needed, it can be re-enabled by setting the
  `create-empty-file: true` configuration in the config file.
- If the application never sends fsync for a file, it will leave behind its
  temp-file (in temp-dir), which will not be cleared until the user unmounts
  the bucket. As an example, if you are writing a large file, and temp-dir
  does not have enough free space available, then you will get 'out of space'
  error. Then the temp-file will not be deleted until you do an fsync for that
  file, or unmount the bucket.

### With Streaming Writes

> Starting with version 2.9.1, GCSFuse supports Streaming writes, which is a new

Reviewer: ```suggestion
Starting with version 2.9.1, GCSFuse supports streaming-writes, which is a new
```

write
path that uploads data directly to Google Cloud Storage (GCS) as it's written
without fully staging the file in the temp-dir. This reduces both latency and
disk space usage, making it particularly beneficial for large, sequential writes
such as checkpoints. Streaming writes can be enabled using
`--enable-streaming-writes` flag or `write:enable-streaming-writes:true` in the
config file.

**Memory Usage:** Each file opened for streaming writes will consume
approximately 64MB of RAM during the upload process. This memory is released
when the file handle is closed. This should be considered when planning resource
allocation for applications using streaming writes.

#### Note on Streaming Writes:

- **New files, Sequential Writes:** Streaming writes are designed for sequential

>   writes to a new, single file only. Modifying existing files, or doing

Author: removed  "single file", it was unnecessary.

  out-of-order writes (whether from the same file handle or concurrent writes
  from multiple file handles) will cause GCSFuse to automatically revert to the
  existing write path of staging writes to a temporary file on disk. An
  informational log message will be emitted when this fallback occurs.

- **Concurrent Writes to the Same File:** While concurrent writes to the same
  file are possible, they are not the primary use case for this initial phase of
  streaming writes. If a (rare, often server-related) error occurs during
  concurrent writes, all file handles must be closed before any future writes
  can resume. This phase of streaming writes is optimized for single-stream

>   writes to new files, , such as for AI/ML checkpointing.

Reviewer: ```suggestion
  writes to new files, such as for AI/ML checkpointing.
```

- **File System Semantics Change:**
    - **FSync Operation does not finalize the Object:** When streaming writes
      are enabled, the fsync operation will not finalize the object on GCS.
      Instead, the object will be finalized only when the file is closed. 
      Only finalized objects are visible to the end user. This is a key 

>       difference from the previous behavior and should be considered when 

Reviewer: ```suggestion
      difference from the default non-streaming-writes behavior and should be considered when 
```

      using streaming writes. Relying on fsync for data durability with
      streaming writes enabled is not recommended. Data is only guaranteed to be
      on GCS after the file is closed.
    - **Rename Operation Syncs the File:** Rename operation on a file undergoing
      writes via streaming writes will be finalized and then renamed. This means
      that any follow up writes will automatically revert to the existing
      behavior of staging writes to a temporary file on disk.
    - **Read Operations During Write:** Today the application can read the data
      when the writes are in progress for that file. With buffered writes, the

>       application will not be able to read until the object is finalized i.e,

Reviewer: ```suggestion
      application will not be able to read the file until the corresponding GCS object is finalized i.e.,
```

  file handles) will cause GCSFuse to automatically revert to the existing write
  path of staging writes to a temporary file on disk. An informational log
  message will be emitted when this fallback occurs.

- **Concurrent Writes to the Same File:** While concurrent writes to the same
  file are possible, they are not the primary use case for this initial phase of
  streaming writes. If a (rare, often server-related) error occurs during
  concurrent writes, all file handles must be closed before any future writes
  can resume. This phase of streaming writes is optimized for single-stream
  writes to new files, such as for AI/ML checkpointing.

- **File System Semantics Change:**
    - **FSync operation does not finalize the object:** When streaming writes
      are enabled, the fsync operation will not finalize the object on GCS.
      Instead, the object will be finalized only when the file is closed. 
      Only finalized objects are visible to the end user. This is a key 
      difference from the default non-streaming-writes behavior and should be considered when 
      using streaming writes. Relying on fsync for data durability with

>       streaming writes enabled is not recommended. Data is only guaranteed to be
>       on GCS after the file is closed.

Reviewer: ```suggestion
      streaming writes enabled is not recommended. Data is guaranteed to be
      on GCS only after the file is closed.
```

    - **Rename Operation Syncs the File:** Rename operation on a file undergoing
      writes via streaming writes will be finalized and then renamed. This means
      that any follow up writes will automatically revert to the existing
      behavior of staging writes to a temporary file on disk.
    - **Read Operations During Write:** Today the application can read the data
      when the writes are in progress for that file. With buffered writes, the
      application will not be able to read the file until the corresponding GCS 

>       object is finalized i.e., fclose() is called. Applications should ensure 
>       that they do not attempt to read from a file while it is being written to 

Reviewer: brevity
```suggestion
      object is finalized i.e., fclose() is called. Applications should not read from a file while it is being written to 
```
